examples/Example13/electrons.cu

```diff
@@ -0,0 +1,341 @@
+          currentTrack.energy, Mass, Charge, geometricalStepLengthFromPhysics, currentTrack.pos, currentTrack.dir,
+          currentTrack.navState, nextState);
+    } else {
+      geometryStepLength =
+          BVHNavigator::ComputeStepAndNextVolume(currentTrack.pos, currentTrack.dir, geometricalStepLengthFromPhysics,
+                                                 currentTrack.navState, nextState, kPush);
+      currentTrack.pos += geometryStepLength * currentTrack.dir;
+    }
+
+    // Set boundary state in navState so the next step and secondaries get the
+    // correct information (currentTrack.navState = nextState only if relocated
+    // in case of a boundary; see below)
+    currentTrack.navState.SetBoundaryState(nextState.IsOnBoundary());
+
+    // Propagate information from geometrical step to MSC.
+    theTrack->SetDirection(currentTrack.dir.x(), currentTrack.dir.y(), currentTrack.dir.z());
+    theTrack->SetGStepLength(geometryStepLength);
+    theTrack->SetOnBoundary(nextState.IsOnBoundary());
+
+    // Apply continuous effects.
+    bool stopped = G4HepEmElectronManager::PerformContinuous(&g4HepEmData, &g4HepEmPars, &elTrack, &rnge);
+
+    // Collect the direction change and displacement by MSC.
+    const double *direction = theTrack->GetDirection();
+    currentTrack.dir.Set(direction[0], direction[1], direction[2]);
+    if (!nextState.IsOnBoundary()) {
+      const double *mscDisplacement = mscData->GetDisplacement();
+      vecgeom::Vector3D<Precision> displacement(mscDisplacement[0], mscDisplacement[1], mscDisplacement[2]);
+      const double dLength2            = displacement.Length2();
+      constexpr double kGeomMinLength  = 5 * copcore::units::nm;          // 0.05 [nm]
+      constexpr double kGeomMinLength2 = kGeomMinLength * kGeomMinLength; // (0.05 [nm])^2
+      if (dLength2 > kGeomMinLength2) {
+        const double dispR = std::sqrt(dLength2);
+        // Estimate safety by subtracting the geometrical step length.
+        safety -= geometryStepLength;
+        constexpr double sFact = 0.99;
+        double reducedSafety   = sFact * safety;
+
+        // Apply displacement, depending on how close we are to a boundary.
+        // 1a. Far away from geometry boundary:
+        if (reducedSafety > 0.0 && dispR <= reducedSafety) {
+          currentTrack.pos += displacement;
+        } else {
+          // Recompute safety.
+          safety        = BVHNavigator::ComputeSafety(currentTrack.pos, currentTrack.navState);
+          reducedSafety = sFact * safety;
+
+          // 1b. Far away from geometry boundary:
+          if (reducedSafety > 0.0 && dispR <= reducedSafety) {
+            currentTrack.pos += displacement;
+            // 2. Push to boundary:
+          } else if (reducedSafety > kGeomMinLength) {
+            currentTrack.pos += displacement * (reducedSafety / dispR);
+          }
+          // 3. Very small safety: do nothing.
```

Comment:
As far as I can read from the code, this branch has also all cases where the displacement exceeds the reduced safety. That means applying MSC depends on the accuracy geometry estimates the safety, at the limit, if geometry returns 0 safety everywhere, there will be no displacement. Shouldn't we try to apply the displacement in all cases, and only check if the boundary gets crossed when we are in this condition? Or maybe I am missing something?

Reply:
All I can say is that this is exactly what Geant4 does in the default case. There are more accurate algorithms for the step limitation, but they won't be implemented in G4HepEm.

Reply:
@mnovak42 any comment on this issue?